### ISSUE TITLE
Identical geometries with floating point coordinates fail `equals?` and `contains?`

### ISSUE THREAD

**Issue by chris** [pkinney/topo#14]
In trying out `equals?` with some box shaped polygons and it doesn't seem to return true, even if I pass it the same shape as seen below. I came across this after having two `Geo.Polygon` items that were the same, but had their points specified in a different order (that also doesn't return true for `equals?`). What's odd is that I tried this with the sample triangle shape in the README and it does return true, so I don't know if it's a box shape bug or what I might be doing wrong?

```
    # This test, passes:
    test "triangle shape int coords satisfies equals" do
      triangle = %Geo.Polygon{coordinates: [[{2, 2}, {20, 2}, {11, 11}, {2, 2}]]}

      assert Topo.contains?(triangle, triangle)
      assert Topo.equals?(triangle, triangle)
    end

    # this test fails
    test "triangle shape float coords satisfies equals" do
      triangle = %Geo.Polygon{coordinates: [[{2.0, 2.0}, {20.0, 2.0}, {11.0, 11.0}, {2.0, 2.0}]]}

      assert Topo.contains?(triangle, triangle)
      assert Topo.equals?(triangle, triangle)
    end
```

Here's the output for example of the failed test:
```
  1) test intersection_of_geometries triangle shape floating point satisfies equals (GreenScoreTest)
     apps/my_score/test/my_score_test.exs:145
     Expected truthy, got false
     code: assert Topo.contains?(triangle, triangle)
     arguments:

         # 1
         %Geo.Polygon{coordinates: [[{2.0, 2.0}, {20.0, 2.0}, {11.0, 11.0}, {2.0, 2.0}]], properties: %{}, srid: nil}

         # 2
         %Geo.Polygon{coordinates: [[{2.0, 2.0}, {20.0, 2.0}, {11.0, 11.0}, {2.0, 2.0}]], properties: %{}, srid: nil}
```

I suspect this is maybe some kind of floating point rounding issue? I noticed that when I made only the first point in the triangle floating point, that it worked properly, but once multiple or all the points were floats, it failed.  This seems problematic in that I'd presume most people's use of GIS values will be floating point (latitude, longitude, etc.). It also makes it hard to write tests that would have the nice feature of using `Topo.equals?` to test two shapes were the same (vs. having to look at individual coordinates one by one - and worse, since the coordinates could be specified in a different order, yet be the same set of coordinates).

Anyway, curious what your thoughts are on this?

**Comment by pkinney** [comment]
@chris Thank you for the very thorough description.

You are right that the equality comparison is not working, but it doesn't actually have to do with rounding error or anything to do with floating-point math specifically.  Rather, when evaluating the Ring/Point relationship, I mistakenly matched only for `{0, 0}` in a function definition and not `{0.0, 0.0}`:

https://github.com/pkinney/topo/blob/2b49e39d22a91e488a4759e0e22f389ee99ba5f2/lib/topo/point_ring.ex#L41

Since here we are change the reference of the ring to put the point at the origin, we look for any vertex of the ring that is at `(0,0)` to know that the point in question equals a vertex. In your case, all of the points should return as `:vertex` since they would each find one vertex at (0.0,0.0), but they were falling through since they were floats and not integers.

https://github.com/pkinney/topo/blob/447e215c944b4be9847877b7870d65ad3d7787a2/lib/topo/point_ring.ex#L41-L42

I went ahead and added an additional match, and your tests now pass.

Additionally, all of the tests within the JTS validation suite used here are integer numbers. In order to catch this sort of issue in the future, I added a "float" copy of each test where the geometries are entered with float coordinates instead of integers.  Sure enough, this showed a slew of failing tests without the change above.  With the change, they are all passing.

I'll push the updated library to Hex later today.
